src/v/cloud_topics/level_one/compaction/committer.cc

```diff
@@ -22,89 +29,403 @@ compaction_committer::compaction_committer(
+    auto id = create_compaction_job_id();
+
+    auto metadata_builder_res
+      = co_await l1::retry_metastore_op_with_default_rtc(
+        [this]() { return _metastore->object_builder(); }, _as);
+    if (!metadata_builder_res.has_value()) {
+        vlog(
+          compaction_log.warn,
+          "Could not create object metadata builder for compaction of tidp {}: "
+          "{}. Aborting.",
+          tidp,
+          metadata_builder_res.error());
+        throw std::runtime_error("Couldn't begin compaction job");
     }
+    auto metadata_builder = std::move(metadata_builder_res).value();
+
+    _compaction_jobs.emplace(
+      id, std::make_unique<job_state>(id, tidp, std::move(metadata_builder)));
+    start_upload_loop(id);
+
+    vlog(
+      compaction_log.debug,
+      "Started compaction job for tidp {} under id {}",
+      tidp,
+      id);
+
+    co_return id;
 }
 
-void compaction_committer::push_update(file_and_md_info update) {
-    _updates.push_back(std::move(update));
-    auto update_response = _policy->on_update(_updates.back());
+void compaction_committer::add_l1_object(
+  compaction_job_id id, file_and_md_info file_and_info) {
+    job_state* job = get_job_by_id(id);
+
+    vassert(
+      job->s == job_state::state::in_progress,
+      "Cannot add l1 object to finalized job.");
+
+    vlog(
+      compaction_log.trace,
+      "Pushing file of size {}, offsets ({}-{}) for compaction of tidp {}, job "
+      "{}",
+      file_and_info.info.size_bytes,
+      file_and_info.ntp_md.base_offset,
+      file_and_info.ntp_md.last_offset,
+      file_and_info.ntp_md.tidp,
+      id);
+
+    auto update_response = _policy->on_update(file_and_info);
+
+    job->staging_file_and_md_infos.push_back(std::move(file_and_info));
+
     if (update_response == committing_policy::update_response::preempt) {
-        _sem.signal();
+        job->upload_sem.signal();
+    }
+}
+
+ss::future<> compaction_committer::finalize_job(
+  compaction_job_id id,
+  chunked_vector<metastore::compaction_update::cleaned_range>
+    new_cleaned_ranges,
+  offset_interval_set removed_tombstone_ranges) {
+    job_state* job = get_job_by_id(id);
+
+    vlog(compaction_log.debug, "Marking job {} as finalized", id);
+    job->s = job_state::state::finalized;
+    job->upload_sem.signal();
+
+    vlog(compaction_log.debug, "Awaiting completion of uploads for job {}", id);
+
+    auto fut = co_await ss::coroutine::as_future(
+      job->all_uploads_complete.wait());
+
+    // Now safe to extract/remove the underlying job state from
+    // `_compaction_jobs`.
+    auto job_ptr_opt = _compaction_jobs.extract(id);
+    vassert(job_ptr_opt.has_value(), "concurrency issue?");
+    auto job_ptr = std::move(job_ptr_opt.value().second);
+
+    finalize_op op = finalize_op::compact_objects;
+    if (fut.failed()) {
+        auto e = fut.get_exception();
+        auto lvl = ssx::is_shutdown_exception(e) ? ss::log_level::debug
```

Comment:
At what point do we stop if we're shutting down? I'm kind of surprised to see us continue on for shutdown errors